scripts/tdr/copy_from_tdr_to_gcs_hca/copy_from_tdr_to_gcs_hca.py

```diff
@@ -0,0 +1,216 @@
+
+
+# for each snapshot get access url and add to a list of access urls for that snapshot
+def get_access_urls(snapshot: str, access_token: str):
+    list_of_access_urls = []
+    # for snapshot in latest_snapshot_ids:
+    logging.info(f"getting access urls for snapshot {snapshot}")
+    files_response = requests.get(
+        f'https://data.terra.bio/api/repository/v1/snapshots/{snapshot}/files?offset=0',
+        headers={'accept': 'application/json', 'Authorization': f'Bearer {access_token}'}
+    )
+    files_response.raise_for_status()
+
+    # Extract file details from the JSON file and add them to a list
+    data = files_response.json()
+    for item in data:
+        list_of_access_urls.append(item['fileDetail']['accessUrl'])
+    return list_of_access_urls
+
+
+def check_staging_is_empty(staging_gs_paths: set[str]):
+    nonempty_staging_areas = []
+    for staging_dir in staging_gs_paths:
+        staging_data_dir = staging_dir + '/data/'
+        logging.info(f'checking contents of staging_data dir: {staging_data_dir}')
+        # using gsutil as output is cleaner & faster
+        output = subprocess.run(['gsutil', 'ls', staging_data_dir], capture_output=True)
+        stdout = output.stdout.strip()
+        files = stdout.decode('utf-8').split('\n')
+        if len(files) > 1:
+            logging.error(f"Staging area {staging_data_dir} is not empty")
+            logging.info(f"files in staging area are: {files}")
+            nonempty_staging_areas.append(staging_data_dir)
+        else:
+            logging.info(f"Staging area {staging_data_dir} is empty")
+
+    if len(nonempty_staging_areas) > 0:
+        logging.error("One or more staging areas are not empty. Exiting.")
+        logging.info(f"Non-empty staging areas are: {nonempty_staging_areas}")
+        sys.exit(1)
+
+
+def copy_tdr_to_staging(tuple_list: list[tuple[str, str]], access_token: str):
+    for project_id in set([x[1] for x in tuple_list]):
+        target_snapshot = f"hca_prod_{project_id.replace('-', '')}"
+        latest_snapshot_id = get_latest_snapshot(target_snapshot, access_token)
+        logging.info(f'latest snapshot id for project {project_id} is {latest_snapshot_id}')
+        access_urls = get_access_urls(latest_snapshot_id, access_token)
+        num_access_urls = len(access_urls)
+        staging_gs_path = [x[0] for x in tuple_list if x[1] == project_id][0]
+        staging_data_dir = staging_gs_path + '/data/'
+        logging.info(f'Copying {num_access_urls} files from snapshot {latest_snapshot_id} to staging area {staging_data_dir}')
+        for access_url in access_urls:
+            # strip the filename from the access url because gcp is not a file system - it's all objects
+            filename = access_url.split('/')[-1]
+            logging.info(f'access_url for snapshot {latest_snapshot_id} is {access_url}')
+            try:
+                subprocess.run(['gcloud', 'storage', 'cp', access_url, staging_data_dir + filename])
```

Comment:
Consider adding a copy manifest to this command, so instead you validating number of files copied (line 187), you can specifically highlight the files not copied successfully.

Reply:
yeah - that's a good idea. But I'm not going to do it this round. I may run this script 2 or 3x a year on 2 or 3 projects, so I'm keeping this as low effort as possible and I've already spent too much time on this. Current effort for value ration is not great.

Reply:
added to possible improvements list in readme